libkbfs/block_retrieval_queue.go

```diff
@@ -163,8 +169,84 @@ func (brq *blockRetrievalQueue) notifyWorker() {
 	}
 }
 
+func (brq *blockRetrievalQueue) CacheAndPrefetch(ptr BlockPointer, block Block,
+	kmd KeyMetadata, priority int, lifetime BlockCacheLifetime,
+	hasPrefetched bool) (err error) {
+	dbc := brq.config.DiskBlockCache()
+	if dbc != nil {
+		go dbc.UpdateMetadata(context.TODO(), ptr.ID)
+	}
+	if hasPrefetched {
+		return brq.config.BlockCache().PutWithPrefetch(ptr, kmd.TlfID(), block,
+			lifetime, true)
+	}
+	if priority < defaultOnDemandRequestPriority {
+		// Only on-demand or higher priority requests can trigger prefetches.
+		return brq.config.BlockCache().PutWithPrefetch(ptr, kmd.TlfID(), block,
+			lifetime, false)
+	}
+	// We must let the cache know at this point that we've prefetched.
+	// 1) To prevent any other Gets from prefetching.
+	// 2) To prevent prefetching if a cache Put fails, since prefetching if
+	//	  only useful when combined with the cache.
+	err = brq.config.BlockCache().PutWithPrefetch(ptr, kmd.TlfID(), block,
+		lifetime, true)
+	if _, isCacheFullError := err.(cachePutCacheFullError); isCacheFullError {
+		brq.log.CDebugf(context.TODO(), "Skipping prefetch because the cache "+
+			"is full")
+		return err
+	}
+	// This must be called in a goroutine to prevent deadlock in case this
+	// CacheAndPrefetch call was triggered by the prefetcher itself.
+	go brq.Prefetcher().PrefetchAfterBlockRetrieved(block, ptr, kmd)
+	return nil
+}
+
+func (brq *blockRetrievalQueue) checkCaches(ctx context.Context,
+	priority int, kmd KeyMetadata, ptr BlockPointer, block Block,
+	lifetime BlockCacheLifetime) error {
+	// Attempt to retrieve the block from the cache. This might be a specific
+	// type where the request blocks are CommonBlocks, but that direction can
+	// Set correctly. The cache will never have CommonBlocks.  TODO: verify
+	// that the returned lifetime here matches `lifetime` (which should always
+	// be TransientEntry, since a PermanentEntry would have been served
+	// directly from the cache elsewhere)?
+	cachedBlock, hasPrefetched, _, err :=
+		brq.config.BlockCache().GetWithPrefetch(ptr)
+	if err == nil && cachedBlock != nil {
+		block.Set(cachedBlock)
+		brq.CacheAndPrefetch(ptr, cachedBlock, kmd, priority, lifetime,
+			hasPrefetched)
+		return nil
+	}
+
+	// Check the disk cache.
+	dbc := brq.config.DiskBlockCache()
+	if dbc == nil {
+		return NoSuchBlockError{ptr.ID}
+	}
+	blockBuf, serverHalf, err := dbc.Get(ctx, kmd.TlfID(), ptr.ID)
+	if err != nil || len(blockBuf) == 0 {
```

Comment:
Do you intend to translate every possible leveldb or disk error into a `NoSuchBlockError`?  I guess that could make sense since the disk cache isn't critical, but I think we should have a comment here about that and log unexpected error types.